code/controllers/subsystem/job.dm

```diff
@@ -454,8 +373,114 @@ SUBSYSTEM_DEF(job)
 			if(!AssignRole(player, SSjob.overflow_role)) //If everything is already filled, make them an assistant
 				return FALSE //Living on the edge, the forced antagonist couldn't be assigned to overflow role (bans, client age) - just reroll
 
+	//Scale number of open security officer slots to population
+	setup_officer_positions()
+
 	return validate_required_jobs(required_jobs)
 
+/datum/controller/subsystem/job/proc/assign_roles(priority = JP_MEDIUM)
+	// Create random orderings for all players
+	var/list/random_orderings = list()
+	// Step 1: Generate random orderings for the players medium jobs
+	for(var/mob/dead/new_player/player in unassigned)
+		var/list/available_jobs = list()
+		// Find all jobs that we are actually able to be
+		for(var/datum/job/job in occupations)
+			if (!is_valid_job(player, job, priority))
+				continue
+			JobDebug("Preparing, Player: [player], Job:[job.title]")
+			available_jobs += job
+		// Create random orderings
+		shuffle_inplace(available_jobs)
+		random_orderings[player] = available_jobs
+		JobDebug("DO [player.ckey] was given the job priority list [jointext(available_jobs, ",")]")
+	// Step 2: Sort the list by the number of availble jobs that each person has, keeping it
+	// random when the amount is the same
+	shuffle_inplace(random_orderings)
+	random_orderings = sortTim(random_orderings, GLOBAL_PROC_REF(cmp_list_size), TRUE)
+	// Step 3: Assign provisional jobs
+	for(var/mob/dead/new_player/player in random_orderings)
+		// Get the first available job for this player
+		for (var/datum/job/job in random_orderings[player])
+			var/job_position_count = job.get_spawn_position_count()
+			if (job.current_positions >= job_position_count && job_position_count != -1)
+				continue
+			// Provisional assignment
+			job.current_positions++
+			player.mind.assigned_role = job.title
+			JobDebug("DO [player.ckey] was assigned the provisional job [job.title]")
+			break
+	// Step 4: Create a random ordering of players
+	// The player list is already shuffled, so we will re-use that for player preference
+	// Step 5: Assign high priority job roles
+	if (priority == JP_MEDIUM)
+		for(var/mob/dead/new_player/player in random_orderings)
+			// Assign high priority jobs
+			for(var/datum/job/job in occupations)
+				if (!is_valid_job(player, job, JP_HIGH))
+					continue
+				var/list/player_job_list = random_orderings[player]
+				// Add this job to the start of the player's preferences list
+				player_job_list.Insert(1, job)
+				JobDebug("DO [player.ckey] requested [job.title] as a high priority job. Updated assignment list: [jointext(player_job_list, ",")]")
+	// Until we reach a stable state or the upper bound is reached, repeatedly
+	// try to get a higher priority role
+	var/iteration_limit = 10
+	var/changed = TRUE
+	while (changed && iteration_limit-- > 0)
+		changed = FALSE
+		for(var/mob/dead/new_player/player in random_orderings)
```

Comment:
This should not use the sorted list, it should be fully random independant of how many job roles you have enabled.

Reply:
Actually we have no way to repeat the medium job assignment using the sorted list, but if it uses the random list then thats fine, because we have already balanced the randomness enough by doing a sorted pass first.